forc-pkg/src/pkg.rs

```diff
@@ -1809,10 +1818,16 @@ pub fn fetch_git(fetch_id: u64, name: &str, pinned: &SourceGitPinned) -> Result<
         // Change HEAD to point to the pinned commit.
         let id = git2::Oid::from_str(&pinned.commit_hash)?;
         repo.set_head_detached(id)?;
-        if path.exists() {
-            let _ = std::fs::remove_dir_all(&path);
-        }
-        std::fs::create_dir_all(&path)?;
+
+        let _ = std::fs::create_dir_all(&path)?;
+        let mut lock = RwLock::new(
+            fs::OpenOptions::new()
+                .write(true)
+                .create(true)
+                .open(path.join(FORC_FILE_LOCK_NAME))?,
```

Comment:
Could we move this expression into a re-usable function? E.g.

```rust
/// Given the path to a directory, lock the directory using a file lock.
///
/// The directory is "unlocked" when the returned `RwLock` is dropped.
///
/// This was introduced to avoid racing on the creation of git checkout directories.
fn lock_dir(dir_path: &Path) -> io::Result<RwLock<File>> {
    let file = fs::OpenOptions::new()
        .write(true)
        .create(true)
        .open(path.join(FORC_FILE_LOCK_NAME))?;
    Ok(RwLock::new(file))
}